cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+        for (request_env, policy_check) in request_envs {
+            // match on the typechecking answer
+            let new_primary_slice = match policy_check {
+                PolicyCheck::Success(typechecked_expr) => {
+                    // compute the trie from the typechecked expr
+                    // using static analysis
+                    compute_root_trie(&typechecked_expr, policy.id())
+                }
+                PolicyCheck::Irrelevant(_) => {
+                    // this policy is ireelevant, so we need no data
+                    Ok(RootAccessTrie::new())
+                }
+
+                // TODO is returning the first error correct?
+                // Also, should we run full validation instead of just
+                // typechecking? Validation does a little more right?
+                PolicyCheck::Fail(errors) => {
+                    // PANIC SAFETY: policy check fail should be a non-empty vector.
+                    #[allow(clippy::expect_used)]
+                    Err(errors
+                        .first()
+                        .expect("Policy check failed without an error")
+                        .clone()
+                        .into())
+                }
+            }?;
+
+            let request_type = request_env
+                .to_request_type()
+                .ok_or(EntityManifestError::PartialRequestError)?;
+            // Add to the manifest based on the request type.
+            if let Some(existing) = manifest.get_mut(&request_type) {
+                *existing = existing.union(&new_primary_slice);
+            } else {
+                manifest.insert(request_type, new_primary_slice);
+            }
+        }
+    }
+
+    Ok(EntityManifest {
+        per_action: manifest,
+    })
+}
+
+/// A static analysis on type-annotated cedar expressions.
+/// Computes the [`RootAccessTrie`] representing all the data required
+/// to evaluate the expression.
+fn compute_root_trie(
+    expr: &Expr<Option<Type>>,
+    policy_id: &PolicyID,
+) -> Result<RootAccessTrie, EntityManifestError> {
+    let mut primary_slice = RootAccessTrie::new();
+    add_to_root_trie(&mut primary_slice, expr, policy_id, false)?;
+    Ok(primary_slice)
+}
+
+/// Add the expression's requested data to the [`RootAccessTrie`].
+/// This handles <expr>s from the grammar (see [`FailedAnalysisError`])
+/// while [`get_expr_path`] handles the <datapath-expr>s.
+fn add_to_root_trie(
+    root_trie: &mut RootAccessTrie,
+    expr: &Expr<Option<Type>>,
+    policy_id: &PolicyID,
+    should_load_all: bool,
+) -> Result<(), EntityManifestError> {
+    match expr.expr_kind() {
+        // Literals, variables, and unkonwns without any GetAttr operations
+        // on them are okay, since no fields need to be loaded.
+        ExprKind::Lit(_) => (),
+        ExprKind::Var(_) => (),
+        ExprKind::Slot(_) => (),
+        ExprKind::Unknown(_) => return Err(EntityManifestError::PartialExpressionError),
```

Comment:
```suggestion
        ExprKind::Lit(_) => Ok(()),
        ExprKind::Var(_) => Ok(()),
        ExprKind::Slot(_) => Ok(()),
        ExprKind::Unknown(_) => Err(EntityManifestError::PartialExpressionError),
```

and then you don't need the separate return of `Ok(())` at the bottom